README.md
# Dynamic motd

The aim of this project is to give some informations when you log into a server through SSH.

Example:

```

   ___  ___ _ ____   _____ _ __
  / __|/ _ \ '__\ \ / / _ \ '__|
  \__ \  __/ |   \ V /  __/ |
  |___/\___|_|    \_/ \___|_|


  Debian GNU/Linux 12 (bookworm) (kernel 6.1.0-7-amd64)


  System information as of Thu Apr 13 08:43:28 2023

  System load:  2.84                 Processes:           338
  Memory usage: 81.69%               Users logged in:     1
  Swap usage:   21.34%
  Disk Usage:
    Usage of /                       : 37.0% of 19.18GB
    Usage of /boot                   : 36.6% of 0.11GB
    Usage of /home                   : 11.1% of 501.60GB
  Inode Usage:
    Usage of /                       : 12.7% of 1286144
    Usage of /boot                   : 1.1% of 31232
    Usage of /home                   : 0.1% of 33406976

  Logged in users:
  user       from laptop.example.org        at Mon Apr  3 09:28:01 2023

No mail.
Last login: Mon Apr  3 07:28:01 2023 from laptop.example.org
```

**Warning** This is Debian and Debian-related distributions only.

## Installation

### Debian package

Go to <https://framagit.org/luc/dynamic-motd/-/releases/permalink/latest>, download the Debian package and its signature file.

Check the signature of the Debian package and install the package:
```bash
minisign -Vm dynamic-motd_*.deb -P RWRzxrp04vb4Db3sle7Az6kSeCipT1ixRjRZPXdUUQuuwgi9UW81E+dx &&
sudo apt install ./dynamic-motd_*.deb
```

### Manual installation

You need to install some packages:

```
apt-get install figlet lsb-release python3-utmp bc
```

Optionnally, you can install `needrestart` which is used to show a message if your server need a reboot (main reason (and the only one I know): you have upgraded your kernel).
If you don't install `needrestart`, it will work, but you won't be warned about the need for a reboot.
`needrestart` warns you about services that need to be restarted too (but is slower than `checkrestart` for that, see below).

You can optionnally install `debian-goodies` which provides `checkrestart`, which will be used to warn you about services that need to be restarted. Relying on `needrestart` for that is slow (±7 seconds) while `checkrestart` do it faster (less than one second).

Then:

```
cp -r update-motd.d/ /etc
rm /etc/motd
ln -s /var/run/motd /etc/motd
```

## Disabling for some users

Just create a `/etc/update-motd.d/hushlogin` file containing the names of the users, like:

```
alice
bob
```

## Salt

You will find a working salt formula in `init.sls`.

```
cd /srv/salt
git clone https://framagit.org/luc/dynamic-motd.git motd
salt your_server state.sls motd
```

## License

GPLv2. Have a look at the [LICENSE file](LICENSE).

## Acknowledments

- Dustin Kirkland, the guy behind the Ubuntu dynamic motd (I took some scripts from Ubuntu and stole inspiration too :D)
- https://github.com/maxis1718/update-motd.d for the skeleton
- https://github.com/jnweiger/landscape-sysinfo-mini for the python script (slightly modified)


CHANGELOG
Revision history for dynamic-motd

0.05 ????-??-??
	- ✨ — Allow to disable the dynamic motd for some users
	- Update salt formula for hushlogin feature
	- Use python3
	- Format scripts with shftmt
	- Use linting on scripts (shellcheck, pylint, mypy)
	- Create a debian package in CI

0.04 2016-08-02
	- Allow to use checkrestart instead of needrestart for services

0.03 2016-07-29
	- Fix bug

0.02 2016-05-04
	- First tagged release


update-motd.d/sysinfo.py
#!/usr/bin/python3
"""
landscape-sysinfo-mini.py -- a trivial re-implementation of the
sysinfo printout shown on debian at boot time. No twisted, no reactor, just /proc & utmp

(C) 2014 jw@owncloud.com

inspired by ubuntu 14.10 /etc/update-motd.d/50-landscape-sysinfo
Requires: python3-utmp 
for counting users.

2014-09-07 V1.0 jw, ad hoc writeup, feature-complete. Probably buggy?
2014-10-08 V1.1 jw, survive without swap
2014-10-13 V1.2 jw, survive without network

Modified by Luc Didry in 2016 and 2023
Get the original version at https://github.com/jnweiger/landscape-sysinfo-mini
"""


import sys
import os
import time
import glob
import utmp  # type: ignore

VERSION = '1.3'

def utmp_count():
    """ Count user processes
    """
    u_r = utmp.UtmpRecord()
    l_users = 0
    for i in u_r:
        if i.ut_type == utmp.USER_PROCESS:
            l_users += 1
    return l_users

def proc_meminfo():
    """ Get memory usage informations
    """
    items = {}
    for line in open('/proc/meminfo', encoding="ASCII").readlines():  # pylint: disable-msg=R1732
        array = line.split()
        items[array[0]] = int(array[1])
    return items

def proc_mount():
    """ Get disks space usage
    """
    items = {}
    for mount in open('/proc/mounts', encoding="ASCII").readlines():  # pylint: disable-msg=R1732
        array = mount.split()
        if array[0].find('/dev/') == 0:
            l_statfs = os.statvfs(array[1])
            perc = 100 - 100. * l_statfs.f_bavail / l_statfs.f_blocks \
                    if l_statfs.f_blocks != 0 else 100
            g_b = l_statfs.f_bsize * l_statfs.f_blocks / 1024. / 1024 / 1024
            items[array[1]] = f"{perc:5.1f}% of {g_b:.2f}GB"
    return items

def inode_proc_mount():
    """ Get disks inode usage
    """
    items = {}
    for mount in open('/proc/mounts', encoding="ASCII").readlines():  # pylint: disable-msg=R1732
        array = mount.split()
        if array[0].find('/dev/') == 0:
            l_statfs = os.statvfs(array[1])
            perc = 100 - 100. * l_statfs.f_ffree / l_statfs.f_files \
                    if l_statfs.f_files != 0 else 100
            i_total = l_statfs.f_files
            items[array[1]] = f"{perc:5.1f}% of {i_total}"
    return items

with open("/proc/loadavg", encoding="ASCII") as avg_line:
    loadav = float(avg_line.read().split()[1])
processes = len(glob.glob('/proc/[0-9]*'))
statfs = proc_mount()
i_statfs = inode_proc_mount()
USERS = utmp_count()
meminfo = proc_meminfo()
memperc = f"{100 - 100. * meminfo['MemAvailable:'] / (meminfo['MemTotal:'] or 1):.2f}%"
SWAPPERC = f"{100 - 100. * meminfo['SwapFree:'] / (meminfo['SwapTotal:'] or 1):.2f}%"

if meminfo['SwapTotal:'] == 0:
    SWAPPERC = '---'

print(f"  System information as of {time.asctime()}\n")
print(f"  System load:  {loadav: <5.2f}                Processes:           {processes}")
print(f"  Memory usage: {memperc: <4}               Users logged in:     {USERS}")
print(f"  Swap usage:   {SWAPPERC}")

print("  Disk Usage:")
for k in sorted(statfs.keys()):
    print(f"    Usage of {k: <24}: {statfs[k]: <20}")

print("  Inode Usage:")
for l in sorted(i_statfs.keys()):
    print(f"    Usage of {l: <24}: {i_statfs[l]: <20}")

if USERS > 0:
    a = utmp.UtmpRecord()
    print("\n  Logged in users:")
    for b in a:
        if b.ut_type == utmp.USER_PROCESS:
            print(f"  \033[1;31m{b.ut_user: <10}\033[m from {b.ut_host: <25}"
                  f" at {time.ctime(b.ut_tv[0]): <20}")
    a.endutent()

sys.exit(0)
